library/ssl_tls.c

```diff
@@ -7351,12 +7347,34 @@ int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
         conf->psk = NULL;
         conf->psk_len = 0;
     }
+
+    /* Remove reference to PSK identity, if any. */
     if( conf->psk_identity != NULL )
     {
         mbedtls_free( conf->psk_identity );
         conf->psk_identity = NULL;
         conf->psk_identity_len = 0;
     }
+}
+
+int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
+                const unsigned char *psk, size_t psk_len,
+                const unsigned char *psk_identity, size_t psk_identity_len )
+{
+    if( psk == NULL || psk_identity == NULL )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    if( psk_len > MBEDTLS_PSK_MAX_LEN )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    /* Identity len will be encoded on two bytes */
+    if( ( psk_identity_len >> 16 ) != 0 ||
```

Comment:
Suggestion for code improvement: if you agree about moving `conf_remove_psk()` to the top, then we might factor out the part about identity (the `NULL` check at the top, this part, allocation + copy below) that's currently a bit duplicated between this function and its opaque variant.

Reply:
Done.